src/map/npc.c

```diff
@@ -1761,7 +1762,22 @@ static bool npc_trader_pay(struct npc_data *nd, struct map_session_data *sd, int
 	npc->trader_ok = false;/* clear */
 
 	snprintf(evname, EVENT_NAME_LENGTH, "%s::OnPayFunds",nd->exname);
-	if ( (ev = strdb_get(npc->ev_db, evname)) ) {
+	if ((ev = strdb_get(npc->ev_db, evname))) {
+		int i, key_nameid = 0, key_amount = 0;
+
+		// Clear the Array
+		script->cleararray_pc(sd, "@bought_nameid", (void*)0);
+		script->cleararray_pc(sd, "@bought_quantity", (void*)0);
+	
+		// save list of bought items
+		for (i = 0; i < VECTOR_LENGTH(*item_list); i++) {
+			struct itemlist_entry *entry = &VECTOR_INDEX(*item_list, i);
+			intptr_t nameid = entry->id;
+			intptr_t amount = entry->amount;
+			script->setarray_pc(sd, "@bought_nameid", i, (void *)nameid, &key_nameid);
```

Comment:
I'm not a fan of char-bound variables, we should rather run the script and inject the variables as scope variables. I'll have to rewrite #1821 though, because I'm not satisfied with the way I implemented it 😕